src/router.rs

```diff
@@ -1,38 +1,129 @@
+        let table = self.table.setup_table(path);
+        let middleware = &*self.middleware;
+        ResourceHandle { table, middleware }
+    }
+
+    /// Add `middleware` to this router.
+    pub fn middleware(&mut self, middleware: impl Middleware<Data> + 'static) -> &mut Self {
+        self.middleware.push(Arc::new(middleware));
+        self
     }
 
     pub(crate) fn route<'a>(
         &'a self,
         path: &'a str,
         method: &http::Method,
-    ) -> Option<(&'a BoxedEndpoint<Data>, RouteMatch<'a>)> {
+    ) -> Option<RouteResult<'a, Data>> {
         let (route, route_match) = self.table.route(path)?;
         // If it is a HTTP HEAD request then check if there is a callback in the endpoints map
         // if not then fallback to the behavior of HTTP GET else proceed as usual
-        if method == http::Method::HEAD && !route.endpoints.contains_key(&http::Method::HEAD) {
-            Some((route.endpoints.get(&http::Method::GET)?, route_match))
-        } else {
-            Some((route.endpoints.get(method)?, route_match))
+        let endpoint =
+            if method == http::Method::HEAD && !route.endpoints.contains_key(&http::Method::HEAD) {
+                route.endpoints.get(&http::Method::GET)?
+            } else {
+                route.endpoints.get(method)?
+            };
+        let middleware = &*route.middleware;
+
+        Some(RouteResult {
+            endpoint,
+            params: route_match,
+            middleware,
+        })
+    }
+}
+
+/// A struct representing specific path of a router.
+///
+/// The struct implements `Deref` and `DerefMut` into `Resource`. You can use this to add a
+/// resource to the path.
+///
+/// Also, using `nest`, you can set up a subrouter for the path.
+pub struct ResourceHandle<'a, Data> {
+    table: &'a mut UrlTable<Resource<Data>>,
+    middleware: &'a [Arc<dyn Middleware<Data> + Send + Sync>],
+}
+
+impl<'a, Data> ResourceHandle<'a, Data> {
+    /// "Nest" a subrouter to the path.
+    pub fn nest<F>(self, builder: F)
+    where
+        F: FnOnce(&mut Router<Data>),
+    {
+        if self.table.resource().is_some() {
+            panic!("This path already has a resource");
         }
+
+        let mut subrouter = Router {
+            table: UrlTable::new(),
+            middleware: self.middleware.to_vec(),
+        };
+        builder(&mut subrouter);
+        std::mem::swap(self.table, &mut subrouter.table);
+    }
+}
+
+impl<'a, Data> std::ops::Deref for ResourceHandle<'a, Data> {
+    type Target = Resource<Data>;
+
+    fn deref(&self) -> &Resource<Data> {
+        self.table
+            .resource()
+            .expect("Resource of this path has not been initialized.")
+    }
+}
+
+impl<'a, Data> std::ops::DerefMut for ResourceHandle<'a, Data> {
+    fn deref_mut(&mut self) -> &mut Resource<Data> {
+        let resource = self.table.resource_mut();
+        if resource.is_none() {
```

Comment:
I think it'd probably work better to *eagerly* initialize the resource when calling `at`. That takes care of the possible panic in deref as well.

If the concern is that we are too eagerly cloning vectors, we could consider putting the middleware list *itself* in an `Arc`, using a copy-on-write-style approach. (We should probably do that regardless)

Reply:
Removed deref things!